iele-syntax.md

```diff
@@ -0,0 +1,232 @@
+  // called using the same syntax as in a local call, e.g
+  //   %pc = call @iele.pc()
+  //   %balance = call @iele.balance(%bank.account)
+  //
+  // The names of the buildins follow the llvm convention for intrinsics: The
+  // name is a valid global name that starts with the prefix "iele.". This means
+  // that no user-defined global name should start with the prefix "iele.".
+  // TODO:The regexp for iele names doesn't reflect that currently, because I
+  // don't know of a way to express that as a regexp and keep the regexp readable.
+
+  syntax IeleName ::= "iele.invalid"     [token]
+  // local state queries
+  syntax IeleName ::= "iele.gas"         [token]
+  syntax IeleName ::= "iele.gasprice"    [token]
+  syntax IeleName ::= "iele.gaslimit"    [token]
+  syntax IeleName ::= "iele.coinbase"    [token]
+  syntax IeleName ::= "iele.timestamp"   [token]
+  syntax IeleName ::= "iele.number"      [token]
+  syntax IeleName ::= "iele.difficulty"  [token]
+  syntax IeleName ::= "iele.address"     [token]
+  syntax IeleName ::= "iele.origin"      [token]
+  syntax IeleName ::= "iele.caller"      [token]
+  syntax IeleName ::= "iele.callvalue"   [token]
+  syntax IeleName ::= "iele.msize"       [token]
+  syntax IeleName ::= "iele.codesize"    [token]
+  syntax IeleName ::= "iele.blockhash"   [token]
+  // account queries
+  syntax IeleName ::= "iele.balance"     [token]
+  syntax IeleName ::= "iele.extcodesize" [token]
+  // precompiled contracts
+  syntax IeleName ::= "iele.ecrec"     [token]
+  syntax IeleName ::= "iele.sha256"    [token]
+  syntax IeleName ::= "iele.rip160"    [token]
+  syntax IeleName ::= "iele.id"        [token]
+  syntax IeleName ::= "iele.ecadd"     [token]
+  syntax IeleName ::= "iele.ecmul"     [token]
+  syntax IeleName ::= "iele.ecpairing" [token]
+
+  // Instructions
+  syntax Instruction ::=
+    AssignInst
+  | LoadInst
+  | StoreInst
+  | SLoadInst
+  | SStoreInst
+  | IsZeroInst
+  | NotInst
+  | AddInst
+  | MulInst
+  | SubInst
+  | DivInst
+  | ExpInst
+  | ModInst
+  | AddModInst
+  | MulModInst
+  | ExpModInst
+  | ByteInst
+  | SExtInst
+  | TwosInst
+  | AndInst
+  | OrInst
+  | XorInst
+  | CmpInst
+  | SHA3Inst
+  | JumpInst
+  | CondJumpInst
+  | LocalCallInst
+  | AccountCallInst
+  | SendInst
+  | ReturnInst
+  | RevertInst
+  | LogInst
+  | CreateInst
+  | SelfdestructInst
+
+  syntax Instructions ::= List{Instruction, ""} [klabel(instructionList)]
+
+  // Blocks
+  syntax LabeledBlock ::= IeleName ":" Instructions
+
+  syntax LabeledBlocks ::= List{LabeledBlock, ""} [klabel(labeledBlockList)]
+
+  syntax UnlabeledBlock ::= Instructions
+
+  syntax Blocks [flatPredicate]
+  syntax Blocks ::= UnlabeledBlock LabeledBlocks | LabeledBlocks
+
+  // Functions
+  syntax FunctionSignature ::= GlobalName "(" FunctionParameters ")"
+
+  syntax FunctionParameters ::= LocalNames
+                              | Int /* when desugared to just the number of parameters */
```

Comment:
This should go into the module where the desugaring happens.  We want the IELE-SYNTAX module to only hold the syntax that can be used for parsing IELE programs.

Reply:
Done!